contributors/design-proposals/loadbalancer-proposal.md

```diff
@@ -0,0 +1,396 @@
+* introduce a new API object `LoadBalancerClaim` to claim a new or existing
+  LoadBalancer (think of PVC)
+* change existing Ingress API to include a `LBSource` field, which either directly
+  uses a LoadBalancer, or via a LoadBalancerClaim (think of Pod using PV or PVC)
+
+We will also introduce network resource such as bandwidth and iops. LoadBalancerClaim
+uses these attributes to claim a LoadBalacer. Though similar to PV/PVC, it is
+important to know that LoadBalancer and LoadBalancerClaim is not 1-to-1 binding;
+rather, it is a 1-to-many relationship. That is, a LoadBalancer can be used to
+serve multiple LoadBalancerClaim.
+
+## Background
+
+### Current Ingress behavior
+
+Ingress can be used to expose a service in the kubernetes cluster:
+
+* cluster admin deploys an ingress-controller Pod beforehand
+* user creates Ingress resource
+* the ingress-controller Pod list&watch **All** Ingress Resources in the cluster,
+  when it sees a new Ingress resource:
+  * on cloud provider, it calls the cloud provider to sync the ingress L7
+    loadbalancing rules
+  * on bare-metal, it syncs nginx (or haproxy, etc) config then reload
+* user out of cluster can then access service in the cluster:
+  * on bare-metal, accessing the node's ip on which ingress-controller Pod is
+    running, ingress-controller Pod will forward request into cluster based on
+    rules defined in Ingress resource
+  * on cloud-provider, accessing the ip provided by cloud provider loadbalancer,
+    cloud provider will forward request into cluster based on rules defined in
+    Ingress Resource
+
+### Limitations of current Ingress implementation
+
+* How many ingress-controller Pods should run in a cluster? Should all
+  ingress-controller Pod list&watch all Ingress resource? There is no way to
+  bind or schedule Ingress resource to a ingress-controller Pod, which result in:
+   * insufficient or excessive use of neworking resource
+   * reload storm when updating Ingress resource, or due to Pod changes
+* Ingress resource is actually internet l7 loadbalancing rules, intranet l7
+  loadbalancing rules has not been supported yet. Eventually, We need a general
+  mechanism for both the Ingress and intranet L7 lb rules "consume" a
+  loadbalancer
+* On bare-metal, it does not provide High Availability because client needs
+  to know the IP addresss of the node where ingress-controller Pod is running.
+  In case of a failure, the ingress-controller Pod will be moved to a different
+  node.
+
+## Goal and NonGoal
+
+### Goal
+
+* Define `LoadBalancer` API
+* Define `LoadBalancerClaim` API
+* Define network attributes of LoadBalancer
+* Define loadbalancer provider, what it is and how it works
+* Define loadbalancer scheduling
+
+### NonGoal
+
+* LoadBalancer HA on bare-metal
+* LoadBalancerClass: different types (internet or intranet), different qos level, etc
+* LoadBalancer scheduling and over-commitment
```

Comment:
Confused -- this seems to appear both in the goal and non-goal sections.

Reply:
Apologies for the misleading. Simple scheduling logic(assign ingress to ingress-controller) is the goal. And  scheduling over-commitment is currently only the goal.